merlin-server/sql/merlin/tables/constraint_specification.sql

```diff
@@ -0,0 +1,28 @@
+create table constraint_specification(
+  plan_id integer not null
+    references plan
+      on update cascade
+      on delete cascade,
+  constraint_id integer not null,
+  constraint_revision integer, -- latest is NULL
+  enabled boolean not null default true,
+
+  constraint constraint_specification_pkey
+    primary key (plan_id, constraint_id),
+  constraint plan_spec_references_existing_constraint
+    foreign key (constraint_id, constraint_revision)
+      references constraint_definition(constraint_id, revision)
+      on update cascade
+      on delete restrict
```

Comment:
Seems like `restrict` could be a bit of a pain if they want to delete a definition that is used in a model spec, and therefore a bunch of plan specs. I'm not sure why it wouldn't be `cascade`. Are you intending that they won't ever need to delete a definition?

Reply:
I am not. Part of the user concerns that lead to versioning was the fear of constraints changing underfoot. Therefore, the decision was made that if a constraint definition is explicitly in use, it cannot be deleted. 

Ie, let's say constraint 10 has 2 definitions, `v1` and `v2`. If the model spec explicitly says to use `v2` (as in, the row on the column is literally `(10, 2)`), then that definition is in use and can't be deleted. If the model spec instead says to use whatever version is the latest (represented in the db as, `(10, NULL)`), then that definition isn't _explicitly_ in use and is fine to delete (from the model's POV, "latest" will become `v1`). 

Though this made me realize that we may need a restriction on the constraints definition table to prevent a constraint from having zero definitions after one's been created... 🤔